compiler/record-hessian/src/HessianComputer.cpp

```diff
@@ -0,0 +1,214 @@
+
+  int out_height = (input_h - dilation_h * (kernel_h - 1) - 1) / stride_h + 1;
+  int out_width = (input_w - dilation_w * (kernel_w - 1) - 1) / stride_w + 1;
+  int patch_size = kernel_h * kernel_w * kernel_ic;
+  std::vector<float> unfolded_buf(input_n * out_height * out_width * patch_size, 0.0f);
+
+  int index = 0;
+  int in_y, in_x;
+  for (int n = 0; n < input_n; ++n)
+  {
+    for (int y = 0; y < out_height; ++y)
+    {
+      for (int x = 0; x < out_width; ++x)
+      {
+        for (int in_c = 0; in_c < input_c; ++in_c)
+        {
+          for (int ky = 0; ky < kernel_h; ++ky)
+          {
+            for (int kx = 0; kx < kernel_w; ++kx)
+            {
+              in_y = y * stride_h + ky * dilation_h;
+              in_x = x * stride_w + kx * dilation_w;
+              if (in_y < input_h && in_x < input_w)
+              {
+                unfolded_buf[index] = buf[((n * input_h + in_y) * input_w + in_x) * input_c + in_c];
+              }
+              index++;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  buf.swap(unfolded_buf);
+}
+
+void HessianComputer::recordHessianForFullyConnected(const luci::CircleNode *node)
+{
+  assert(_input_tensor->shape().rank() < 4);
+  assert(_input_tensor->element_type() == DataType::FLOAT32);
+
+  uint32_t size_in_ch;
+  uint32_t length;
+
+  const auto data = _input_tensor->data<float>();
+  const auto num_elements = _input_tensor->shape().num_elements();
+  std::vector<float> buf(data, data + num_elements);
+
+  if (_input_tensor->shape().num_dims() == 3)
+  {
+    size_in_ch = _input_tensor->shape().dim(2); // input_tensor [batch, length, channel]
+  }
+  else if (_input_tensor->shape().num_dims() == 2)
+  {
+    size_in_ch = _input_tensor->shape().dim(1); // input_tensor [length, channel]
+  }
+  else
+  {
+    throw std::runtime_error("RecordHessian: Unsupported node rank");
+  }
+  assert(size_in_ch != 0);
+  length = num_elements / size_in_ch;
+
+  std::vector<float> hessian(size_in_ch * size_in_ch, 0);
+
+  for (int i = 0; i < size_in_ch; ++i)
+  {
+    for (int j = 0; j < size_in_ch; ++j)
+    {
+      float sum = 0;
+      for (int k = 0; k < length; ++k)
+      {
+        sum += buf[i + k * size_in_ch] * buf[j + k * size_in_ch];
+      }
+      hessian[i * size_in_ch + j] = 2 * sum;
+    }
+  }
+
+  HessianVector &vector = _hessian_map[node];
+  vector.update(hessian);
+}
+
+void HessianComputer::recordHessianForConv2D(const luci::CircleNode *node)
+{
+  assert(_input_tensor->shape().rank() == 4);
+  assert(_input_tensor->element_type() == DataType::FLOAT32);
+
+  const auto circle_conv2d = loco::must_cast<const luci::CircleConv2D *>(node);
+  const auto node_filter = loco::must_cast<luci::CircleConst *>((circle_conv2d)->filter());
+  assert(node_filter.dtype() == loco::DataType::FLOAT32);
+
+  uint32_t size_filter = node_filter->size<loco::DataType::FLOAT32>();
+  uint32_t size_in_ch = node_filter->size<loco::DataType::FLOAT32>() / node->dim(3).value();
```

Comment:
it would be better to use `circle_conv2d` instead of `node`

there is no conditional check if `circle_conv2d` (or `node`) `rank >= 4`